### PR TITLE
Report root cause when a stack update fails

### PR DESCRIPTION
**Pull request by deorus**
Addresses Doist/platform-backlog#734 — failed deploys currently surface only a single, often-wrong error line, leaving you to dig through the AWS Console. This action is the standard deploy step across ~30 Doist repos, so the improvement applies broadly, not just to Todoist.

## What changed
- On failure, scan the stack events for this update's `ClientRequestToken` and collect **all** genuinely-failed resources (any `*_FAILED` status, skipping cancellation cascades). Report the **earliest** failure as the root cause — the previous heuristic captured the *newest* failed event, which is usually a downstream symptom.
- Under GitHub Actions, write a summary table to the job summary: Time (UTC) / Resource / Type / Status / Reason, plus a deep link to the stack's events tab in the AWS Console. Remaining failures are emitted as `::error::` annotations.
- Falls back to a generic message + console link when no resource-level reason is available.

## Notes
- No new IAM permission — `cloudformation:DescribeStackEvents` was already required.
- Tests cover failure detection, root-cause ordering, Markdown-cell escaping, and the console URL.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
-					case types.ResourceStatusUpdateRollbackComplete,
-						types.ResourceStatusRollbackFailed:
-						return cmp.Or(likelyRootCause, fmt.Errorf("%v, see AWS CloudFormation Console for more details", evt.ResourceStatus))
 					case types.ResourceStatusUpdateComplete:
 						return nil
+					case types.ResourceStatusUpdateRollbackComplete,
+						types.ResourceStatusRollbackFailed:
+						terminal = evt.ResourceStatus
 					}
+					continue
```
Comment (doistbot) [inline main.go:144]: <big><kbd>🟠 P1</kbd></big>  The `continue` statement skips all stack-level events from being evaluated by `isFailure`. While this hides the redundant "The following resource(s) failed" stack event during typical deployments, it also drops stack-level pre-flight failures (e.g., missing IAM capabilities, export overlaps, or unresolvable parameters) where no underlying resources fail. This results in the action returning a generic `UPDATE_ROLLBACK_COMPLETE` error and dropping the actual root cause, which is a regression from the previous behavior.

Remove `continue` so that stack-level `_FAILED` events are still collected. Because stack-level events occur after resource failures, `sortedFailures` will still correctly place the true resource root cause first when both exist.

File: main.go [hunk 2]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+		}
+	}
+	if len(failures) == 0 {
+		return fmt.Errorf("%s, see stack events: %s", terminal, consoleURL)
+	}
+	for _, e := range failures[1:] {
+		log.Printf("%s%s (%s) %s: %s", githubErrPrefix, e.logicalID, e.resType, e.status, oneLine(e.reason))
+	}
+	root := failures[0]
+	return fmt.Errorf("%s (%s) %s: %s — see stack events: %s", root.logicalID, root.resType, root.status, oneLine(root.reason), consoleURL)
```
Comment (doistbot) [inline main.go:206]: <big><kbd>🟡 P2</kbd></big>  `ResourceStatusReason` is optional in CloudFormation events, so a `*_FAILED` event can legitimately arrive with no reason. In that case this now returns `Resource (Type) STATUS:  — see stack events`, even though the PR description says to fall back to a generic message when no resource-level reason is available. Please treat an empty `root.reason` as missing here and either omit the `: %s` fragment or fall back to the terminal-status message + console link.

File: main.go [hunk 3]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+}
+
+func eventsConsoleURL(region, stackID string) string {
+	return fmt.Sprintf("https://%[1]s.console.aws.amazon.com/cloudformation/home?region=%[1]s#/stacks/events?stackId=%s",
```
Comment (doistbot) [inline main.go:210]: <big><kbd>🟡 P2</kbd></big>  This hardcodes the commercial AWS Console domain, so the deep link is wrong for GovCloud and China regions even though the action itself can run there. Please derive the console host from the stack ARN/partition (or otherwise map the region to the correct console domain) instead of always using `*.console.aws.amazon.com`.

File: main.go [hunk 4]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+				}
+				if isFailure(evt.ResourceStatus, unptr(evt.ResourceStatusReason)) {
+					fe := failedEvent{
+						logicalID: unptr(evt.LogicalResourceId),
+						resType:   unptr(evt.ResourceType),
+						status:    evt.ResourceStatus,
+						reason:    unptr(evt.ResourceStatusReason),
+						timestamp: unptr(evt.Timestamp),
+					}
+					failures[fe.logicalID+"\x00"+fe.timestamp.String()] = fe
```
Comment (doistbot) [inline main.go:154]: <big><kbd>🟡 P2</kbd></big>  Replace this synthetic composite key with the event's unique ID (`unptr(evt.EventId)`). CloudFormation guarantees a stable `EventId` for every stack event, which provides a safer, allocation-free way to deduplicate events without relying on stringified timestamps.

File: main.go [hunk 5]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+func isFailure(status types.ResourceStatus, reason string) bool {
+	if !strings.HasSuffix(string(status), "_FAILED") {
+		return false
+	}
+	return !strings.Contains(strings.ToLower(reason), "cancelled")
+}
+
+func sortedFailures(m map[string]failedEvent) []failedEvent {
+	out := slices.Collect(maps.Values(m))
+	slices.SortFunc(out, func(a, b failedEvent) int { return a.timestamp.Compare(b.timestamp) })
```
Comment (doistbot) [inline main.go:184]: <big><kbd>🟡 P2</kbd></big>  This only sorts by timestamp, but the failures were first collected into a map, so any two events with the same timestamp end up in arbitrary order. Since `reportFailure()` treats `failures[0]` as the root cause, the reported root failure can change run to run for equal-timestamp events. Keep scan order in a slice (using a separate set only for dedupe if needed), or add a deterministic tiebreaker before selecting the first failure.

File: main.go [hunk 6]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
-				}
 				debugf("%s\t%s\t%v", unptr(evt.ResourceType), unptr(evt.LogicalResourceId), evt.ResourceStatus)
 				if unptr(evt.LogicalResourceId) == stackName && unptr(evt.ResourceType) == "AWS::CloudFormation::Stack" {
 					switch evt.ResourceStatus {
-					case types.ResourceStatusUpdateRollbackComplete,
-						types.ResourceStatusRollbackFailed:
-						return cmp.Or(likelyRootCause, fmt.Errorf("%v, see AWS CloudFormation Console for more details", evt.ResourceStatus))
 					case types.ResourceStatusUpdateComplete:
 						return nil
+					case types.ResourceStatusUpdateRollbackComplete,
```
Comment (doistbot) [inline main.go:140]: <big><kbd>🟡 P2</kbd></big>  This change stops returning immediately on rollback and now keeps paginating `DescribeStackEvents` until the one-hour cutoff. On stacks with a lot of recent history, that means extra API pages for unrelated older operations right on the failure path. Since events are newest-first, consider breaking once you've reached the current update's start event for this `ClientRequestToken` (for example the matching stack `UPDATE_IN_PROGRESS` event) instead of scanning the whole cutoff window.

File: main.go [hunk 7]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+	out := slices.Collect(maps.Values(m))
+	slices.SortFunc(out, func(a, b failedEvent) int { return a.timestamp.Compare(b.timestamp) })
+	return out
+}
+
+// reportFailure writes a human-readable failure report — a GitHub step summary
+// table plus per-resource error annotations — and returns the most likely root
+// cause as an error. The earliest failure is the root cause; later ones usually
+// cascade from it.
+func reportFailure(region, stackID string, terminal types.ResourceStatus, failures []failedEvent) error {
```
Comment (doistbot) [inline main.go:192]: <big><kbd>🟡 P2</kbd></big>  reportFailure is the core of this PR — it picks failures[0] as the root cause, formats the error message, logs failures[1:] as annotations, and falls back to a generic message when there are no resource-level failures. None of this is tested. A regression here (wrong element selected, broken fallback, malformed message) would directly reintroduce the wrong-root-cause problem this PR fixes. It's a pure function in the same package — with GITHUB_STEP_SUMMARY unset, a test can pass a []failedEvent and assert the returned error string for both the populated and empty cases.

File: main.go [hunk 8]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+	root := failures[0]
+	return fmt.Errorf("%s (%s) %s: %s — see stack events: %s", root.logicalID, root.resType, root.status, oneLine(root.reason), consoleURL)
+}
+
+func eventsConsoleURL(region, stackID string) string {
+	return fmt.Sprintf("https://%[1]s.console.aws.amazon.com/cloudformation/home?region=%[1]s#/stacks/events?stackId=%s",
+		region, url.QueryEscape(stackID))
+}
+
+func writeStepSummary(path, consoleURL string, terminal types.ResourceStatus, failures []failedEvent) error {
```
Comment (doistbot) [inline main.go:214]: <big><kbd>🟡 P2</kbd></big>  writeStepSummary produces the primary new user-facing output — the Markdown table written to ~30 repos' job summaries. The table header, separator row, per-row timestamp formatting, and console link assembly are all untested logic that mdCell/eventsConsoleURL tests alone don't cover. It's testable with a temp file: call writeStepSummary with the temp path, read it back, and assert the rendered Markdown. A regression would produce garbled summaries across all consuming repos.

**Review by doistbot — COMMENTED** [review]
This PR improves CloudFormation deploy failure reporting by scanning stack events for the update's ClientRequestToken, identifying the earliest failed resource as the root cause, and writing a summary table with console links to the GitHub Actions job summary.

Few things worth tightening:
- **Stack-level failures get dropped:** The `continue` statement that skips stack events also hides pre-flight failures (missing IAM capabilities, export overlaps, etc.) where no resource fails, resulting in a generic error instead of the real cause. Remove the `continue` so stack-level `_FAILED` events are still collected — since stack events are newer than resource failures, `sortedFailures` will still place the true root cause first.
- **Edge cases in the root-cause message:** A `*_FAILED` event with no `ResourceStatusReason` produces a malformed message (`… STATUS: — see stack events`). Treat an empty `root.reason` as missing and fall back to the generic message + console link. Separately, the console URL hardcodes `console.aws.amazon.com`, which is wrong for GovCloud and China partitions — derive the host from the stack ARN/partition instead.
- **Nondeterministic root-cause selection:** Events collected into a map then sorted only by timestamp can shuffle equal-timestamp failures, making the reported root cause vary run to run. Use `EventId` for dedup instead of the composite key, and add a deterministic tiebreaker or preserve scan order so `failures[0]` is stable.
- **Polling and testing gaps:** On busy stacks, pagination now continues through the full one-hour cutoff even after rollback — break early once the matching `UPDATE_IN_PROGRESS` event for this token is found. Also, the two highest-value functions (`reportFailure` and `writeStepSummary`) are untested despite being pure functions with an existing test harness; adding coverage for both the populated and empty cases would guard against regressions in the core logic this PR introduces.

I also included a few optional follow-up notes in the details below.

<details>
<summary>Optional follow-up note (1)</summary>

- **[P3] main.go:146**: Since the stack's terminal event is always newer than the resource failures that caused it, it will be encountered first in this newest-to-oldest stream. You can avoid redundantly populating the `failures` map on every tick during a rollback by checking `if terminal != "" && isFailure(...)`. This safely defers all map operations and string parsing until the final tick when the map is actually used.

</details>

<sub>[Share Feedback](https://github.com/Doist/doistbot-app/issues/new?template=review-feedback.yml&title=Review+feedback+for+Doist%2Fupdate-cloudformation-stack&pr_url=https%3A%2F%2Fgithub.com%2FDoist%2Fupdate-cloudformation-stack%2Fpull%2F3&datadog_url=https%3A%2F%2Fapp.datadoghq.com%2Flogs%3Fquery%3Dservice%253A%2528doistbot-app-sandbox%2520OR%2520doistbot-app%2529%2520%2540meta.deliveryId%253A959640a0-6e05-11f1-9800-923c4d26959e%26agg_m%3Dcount%26agg_m_source%3Dbase%26agg_t%3Dcount%26clustering_pattern_field_path%3Dmessage%26cols%3Dhost%252Cservice%26fromUser%3Dtrue%26messageDisplay%3Dinline%26refresh_mode%3Dsliding%26storage%3Dhot%26stream_sort%3Ddesc%26viz%3Dstream%26live%3Dtrue) • [Review Logs](https://app.datadoghq.com/logs?query=service%3A%28doistbot-app-sandbox%20OR%20doistbot-app%29%20%40meta.deliveryId%3A959640a0-6e05-11f1-9800-923c4d26959e&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=host%2Cservice&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&live=true)</sub>

File: main.go [hunk 9]
```diff
@@ -105,14 +106,18 @@ func run(ctx context.Context, stackName string, args []string) error {
...
 	for {
 		select {
 		case <-ticker.C:
 		case <-ctx.Done():
 			return ctx.Err()
 		}
+		// Re-scan events on every tick. The stack-level terminal event is the
+		// newest one, but the resource failures that caused it are older, so on
+		// failure we keep scanning the page to collect them all before reporting.
+		failures := make(map[string]failedEvent)
```
Comment (artyom) [inline main.go:119]: You don't need a map here, `[]failedEvent` is enough.

File: main.go [hunk 10]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
-						return cmp.Or(likelyRootCause, fmt.Errorf("%v, see AWS CloudFormation Console for more details", evt.ResourceStatus))
 					case types.ResourceStatusUpdateComplete:
 						return nil
+					case types.ResourceStatusUpdateRollbackComplete,
+						types.ResourceStatusRollbackFailed:
+						terminal = evt.ResourceStatus
 					}
+					continue
+				}
+				if isFailure(evt.ResourceStatus, unptr(evt.ResourceStatusReason)) {
```
Comment (artyom) [inline main.go:146]: No reason to have a dedicated function since this is a single place where it's used. It can be inlined and still be a one-line condition.

File: main.go (hunk 4 above)
Comment (artyom) [reply main.go:154]: Decent observation, but there's no need to have a map *at all*, and llms missed this.

File: main.go (hunk 3 above)
Comment (artyom) [inline main.go:210]: You don't need the region at all.

```go
func consoleURL(arn string) string {
  return "https://console.aws.amazon.com/go/view?arn="+url.QueryEscape(arn)
}
```

File: main.go [hunk 11]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+	return out
+}
+
+// reportFailure writes a human-readable failure report — a GitHub step summary
+// table plus per-resource error annotations — and returns the most likely root
+// cause as an error. The earliest failure is the root cause; later ones usually
+// cascade from it.
+func reportFailure(region, stackID string, terminal types.ResourceStatus, failures []failedEvent) error {
+	consoleURL := eventsConsoleURL(region, stackID)
+	if path := os.Getenv("GITHUB_STEP_SUMMARY"); path != "" {
```
Comment (artyom) [inline main.go:194]: Bad pattern/code smell — functions should now have side effects like this.

Instead, reportFailure can return a dedicated error type that has a specific method taking a file name as the argument to write report to.

File: main.go (hunk 8 above)
Comment (artyom) [inline main.go:214]: This can be a method on a dedicated error type, and only take a single argument — file name where to write report to.

File: main.go [hunk 12]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+	return fmt.Errorf("%s (%s) %s: %s — see stack events: %s", root.logicalID, root.resType, root.status, oneLine(root.reason), consoleURL)
+}
+
+func eventsConsoleURL(region, stackID string) string {
+	return fmt.Sprintf("https://%[1]s.console.aws.amazon.com/cloudformation/home?region=%[1]s#/stacks/events?stackId=%s",
+		region, url.QueryEscape(stackID))
+}
+
+func writeStepSummary(path, consoleURL string, terminal types.ResourceStatus, failures []failedEvent) error {
+	f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0o644)
```
Comment (artyom) [inline main.go:215]: You can just build the content into bytes.Buffer instead of strings.Builder as the code already does, and then write it like this:

```
return os.WriteFile(filename, b.Bytes(), 0666)
```

[the github step summary is unique for each step](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-commands#:~:text=GITHUB_STEP_SUMMARY%20is%20unique%20for%20each%20step%20in%20a%20job.), so you can just write it without being afraid about overwriting it.

File: main.go [hunk 13]
```diff
@@ -127,25 +132,112 @@ func run(ctx context.Context, stackName string, args []string) error {
...
+		b.WriteString("\n")
+	}
+	fmt.Fprintf(&b, "[View stack events in the AWS Console](%s)\n", consoleURL)
+	_, err = io.WriteString(f, b.String())
+	return err
+}
+
+// oneLine collapses runs of whitespace (including newlines) into single spaces
+// so a reason renders on a single log line or table cell.
+func oneLine(s string) string { return strings.Join(strings.Fields(s), " ") }
```
Comment (artyom) [inline main.go:236]: A more efficient way to do this would be:

```go
func oneLine(s string) string {
	var prevIsSpace bool
	f := func(r rune) rune {
		switch {
		case unicode.IsSpace(r):
			if prevIsSpace {
				return -1
			}
			prevIsSpace = true
			return ' '
		default:
			prevIsSpace = false
			return r
		}
	}
	return strings.TrimSpace(strings.Map(f, s))
}
```

File: main_test.go [hunk 14]
```diff
@@ -27,3 +32,56 @@ func Test_parseKvs(t *testing.T) {
 		}
 	}
 }
+
+func Test_isFailure(t *testing.T) {
```
Comment (artyom) [inline main_test.go:36]: Useless test, can be dropped.

File: main_test.go [hunk 15]
```diff
@@ -27,3 +32,56 @@ func Test_parseKvs(t *testing.T) {
...
+	}
+}
+
+func Test_mdCell(t *testing.T) {
+	if got, want := mdCell("a | b\nc"), `a \| b c`; got != want {
+		t.Errorf("mdCell() = %q, want %q", got, want)
+	}
+}
+
+func Test_eventsConsoleURL(t *testing.T) {
```
Comment (artyom) [inline main_test.go:81]: Useless test.

**Review by artyom — CHANGES_REQUESTED** [review]

**Comment by artyom** [comment]
Next time let's maybe briefly discuss the approach in a github issue first before unleashing LLMs on the task, otherwise it takes more time to understand what they did. 😅

File: main.go (hunk 9 above)
Comment (deorus) [reply main.go]: Done, plain slice now. Events come newest first, so I append in scan order and reverse once at the end. No sorting or dedupe needed.

File: main.go (hunk 10 above)
Comment (deorus) [reply main.go]: Inlined.

File: main.go (hunk 4 above)
Comment (deorus) [reply main.go]: Yep, gone along with the map.

File: main.go (hunk 3 above)
Comment (deorus) [reply main.go]: Took your link, region is gone from the whole path.

File: main.go (hunk 11 above)
Comment (deorus) [reply main.go]: `run` now returns `*updateFailed` (stack ARN, terminal status, failures oldest first). `Error()` gives the root cause line, `writeSummary(filename)` renders the table, and `main` does the `errors.As` and writes to `GITHUB_STEP_SUMMARY` when set. I also dropped the extra `::error::` annotations for the non-root failures since the summary table already lists them.

File: main.go (hunk 8 above)
Comment (deorus) [reply main.go]: Done, it's a method on the error type taking only the file name.

File: main.go (hunk 12 above)
Comment (deorus) [reply main.go]: Done, thanks for the pointer on the per-step summary file.

File: main.go (hunk 13 above)
Comment (deorus) [reply main.go]: Took your version.

File: main_test.go (hunk 14 above)
Comment (deorus) [reply main_test.go]: Dropped. Replaced with one test for the error type: `Error()` with and without failures, and the rendered summary via a temp file.

File: main_test.go (hunk 15 above)
Comment (deorus) [reply main_test.go]: Dropped.

File: main.go (hunk 1 above)
Comment (deorus) [reply main.go]: Removed the `continue`, stack-level failed events go into the same list now. Missing capabilities and export clashes are rejected by the UpdateStack call itself though, so they never reached this loop.

File: main.go (hunk 2 above)
Comment (deorus) [reply main.go]: Handled, `Error()` omits the reason fragment when it's empty.

File: main.go (hunk 3 above)
Comment (deorus) [reply main.go]: Superseded by the ARN-only `/go/view` link, no region or partition mapping needed.

File: main.go (hunk 5 above)
Comment (deorus) [reply main.go]: Moot, there's no map and no sort anymore. Order is the scan order from the API.

File: main.go (hunk 6 above)
Comment (deorus) [reply main.go]: Added, the scan breaks at this token's stack `UPDATE_IN_PROGRESS` event.

File: main.go (hunk 7 above)
Comment (deorus) [reply main.go]: Covered by `Test_updateFailed`.

File: main.go (hunk 8 above)
Comment (deorus) [reply main.go]: Covered by `Test_updateFailed`.

**Comment by deorus** [comment]
Fair, will open an issue first next time.
